nav2_route/src/plugins/edge_cost_functions/start_pose_orientation_scorer.cpp

```
@@ -0,0 +1,95 @@
+  const std::string & name)
+{
+  RCLCPP_INFO(node->get_logger(), "Configuring start pose orientation scorer.");
+  name_ = name;
+  logger_ = node->get_logger();
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    getName() + ".orientation_tolerance", rclcpp::ParameterValue(M_PI / 2.0));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node, getName() + ".orientation_weight", rclcpp::ParameterValue(1.0));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node, getName() + ".use_orientation_threshold", rclcpp::ParameterValue(false));
+
+  orientation_tolerance_ = node->get_parameter(getName() + ".orientation_tolerance").as_double();
+  orientation_weight_ =
+    static_cast<float>(node->get_parameter(getName() + ".orientation_weight").as_double());
+  route_frame_ = node->get_parameter("route_frame").as_string();
+  base_frame_ = node->get_parameter("base_frame").as_string();
+  use_orientation_threshold_ =
+    node->get_parameter(getName() + ".use_orientation_threshold").as_bool();
+
+  tf_buffer_ = tf_buffer;
+}
+
+bool StartPoseOrientationScorer::score(
+  const EdgePtr edge,
+  const geometry_msgs::msg::PoseStamped & /* goal_pose */,
+  const EdgeType & edge_type, float & cost)
+{
+  if (edge_type == EdgeType::START) {
+    double edge_orientation = std::atan2(
+      edge->end->coords.y - edge->start->coords.y,
+      edge->end->coords.x - edge->start->coords.x);
+    geometry_msgs::msg::PoseStamped robot_pose = getRobotPose();
+    double start_orientation = tf2::getYaw(robot_pose.pose.orientation);
+    double d_yaw = std::abs(angles::shortest_angular_distance(edge_orientation, start_orientation));
+
+    if (use_orientation_threshold_) {
+      if (d_yaw > orientation_tolerance_) {
+        return false;
+      }
+    }
+
+    cost = orientation_weight_ * static_cast<float>(d_yaw);
+  }
+  return true;
+}
+
+geometry_msgs::msg::PoseStamped StartPoseOrientationScorer::getRobotPose()
```

Teh goal intent extractor has the start pose as either the action message's start or the current pose from TF depending on the request settings, in `findStartandGoal`. I think we should get the start pose that we should use in a new API `getStart` and pass that through the server to the scoring plugins with the goal pose. That way we're using the proper start pose that the user requested. Also, TF lookups like this each `score()` edge would be very computationally expensive, we would want to compute that once and cache it instead anyway. 

Similarly to the goal orientation scorer, if `== PoseStamped()` because its not used as we're working with the Node ID search not poses, then throw exception as an invalid plugin to use for the request type. 

Also, add in the docstrings for both of these plugins that its only valid when used with pose-request types (not node request types)

---

I think that would also remove the need to obtain the route and base frames

---

modified all the edge scorers and this one in particular, to use the `RouteData` struct that contains all the necessary information. As you've recommended I've populated the `start_pose` field in the `RouteData` struct with the newly added `getStart()` goal intent extractor method. Deleted a few of the irrelevant parameters and the `getRobotPose()` method